File: web-semantic-explorer/frontend/src/shared/api/taxonomy.ts
import type { CancelablePromise } from "@/client/core/CancelablePromise"
import { OpenAPI } from "@/client/core/OpenAPI"
import { request } from "@/client/core/request"

export type AuthorOption = {
  name: string
}

export type AuthorsListResponse = {
  authors: AuthorOption[]
}

export type CategoryOption = {
  name: string
}

export type CategoriesListResponse = {
  categories: CategoryOption[]
}

export function listAuthors(): CancelablePromise<AuthorsListResponse> {
  return request(OpenAPI, {
    method: "GET",
    url: "/api/v1/taxonomy/authors",
  })
}

export function listCategories(): CancelablePromise<CategoriesListResponse> {
  return request(OpenAPI, {
    method: "GET",
    url: "/api/v1/taxonomy/categories",
  })
}

File: web-semantic-explorer/backend/app/api/routes/taxonomy.py
from fastapi import APIRouter
from sqlmodel import select

from app.api.deps import SessionDep
from app.models.relations import ArticleAuthor, ArticleCategory
from app.models.taxonomy import Author, Category
from app.schemas.taxonomy import (
    AuthorOption,
    AuthorsListResponse,
    CategoriesListResponse,
    CategoryOption,
)

router = APIRouter(prefix="/taxonomy", tags=["taxonomy"])


@router.get("/authors", response_model=AuthorsListResponse)
def list_authors(session: SessionDep) -> AuthorsListResponse:
    """Autores con al menos un artículo, ordenados por nombre (para filtros exactos)."""
    rows = session.exec(
        select(Author.name)
        .join(ArticleAuthor, ArticleAuthor.author_id == Author.id)
        .distinct()
        .order_by(Author.name)
    ).all()

    return AuthorsListResponse(
        authors=[AuthorOption(name=name) for name in rows if name]
    )


@router.get("/categories", response_model=CategoriesListResponse)
def list_categories(session: SessionDep) -> CategoriesListResponse:
    """Categorías con al menos un artículo, ordenadas por nombre (para filtros exactos)."""
    rows = session.exec(
        select(Category.name)
        .join(ArticleCategory, ArticleCategory.category_id == Category.id)
        .distinct()
        .order_by(Category.name)
    ).all()

    return CategoriesListResponse(
        categories=[CategoryOption(name=name) for name in rows if name]
    )
